### ISSUE TITLE
Model generates an extra box (prediction) on top of correct box (prediction)

### ISSUE THREAD

**Issue by GeoTox** [ultralytics/yolov5#3580]
## ❔Question
Hey guys, I'm trying to train yolov5m on license plate data. Although it works well, it sometimes behaves strangely. In most of the incorrect prediction cases it just adds an extra box/prediction on top of the one it predicted correctly. The incorrect prediction label repeats the label of a previous prediction. Here are examples for you to see yourselves.
[Image #1](https://i.imgur.com/LASeCiY.png) The box with the label '5' was placed on top of the correctly predicted box with label '0'.
[Image #2](https://i.imgur.com/MZKrGai.png) 'S' over 'W'
[Image #3](https://i.imgur.com/IR3F8k4.png) 'U' over 'V'
Might this be because of a bug or something or is it a normal behaviour of yolov5 models?

## Additional context
yolov5m was trained on 200k images for 300 epochs. Images were already augmented, so yolo augmentations were set to be off.

**Comment by glenn-jocher** [comment]
@GeoTox this is multi-label detection, which is enabled by default when training but has been disabled by default in detect.py. I would recommend you update your code to the latest master first. To update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload with `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

More information on multi-label detection differences between [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py), [detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py) and [test.py](https://github.com/ultralytics/yolov5/blob/master/test.py) in YOLOv5.

These 3 files are designed for different purposes and utilize different dataloaders with different settings. train.py dataloaders are designed for a speed-accuracy compromise, test.py is designed to obtain the best mAP on a validation dataset, and detect.py is designed for best real-world inference results. A few important aspects of each:

## `train.py`

- **trainloader** LoadImagesAndLabels(): designed to load train dataset images and labels. Augmentation capable and enabled. https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/train.py#L188-L192
- **testloader** LoadImagesAndLabels(): designed to load val dataset images and labels. Augmentation capable but disabled. https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/train.py#L199-L202
- image size: `640`
- rectangular inference: `False`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): `0.001`
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): `0.6`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): `True`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): `None`

## `test.py`

- **dataloader** LoadImagesAndLabels(): designed to load train, val, test dataset images and labels. Augmentation capable but disabled. https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/test.py#L89-L90
- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): `0.001`
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): `0.6`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): `True`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): `0.5 * maximum stride`

## `detect.py`

- **dataloaders** (multiple): designed for loading multiple types of media (images, videos, globs, directories, streams). https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/detect.py#L46-L53
- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L152): `0.25` 
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L153): `0.45`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/utils/general.py#L415): `False`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L45-L54): `None`

## YOLOv5 PyTorch Hub

`models.autoShape()` class used for image loading, preprocessing, inference and NMS. For more info see [YOLOv5 PyTorch Hub Tutorial](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading) https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/models/common.py#L225-L250

- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `0.25 `
- [iou threshold](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `0.45`
- [multi-label](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L288): `False`
- [padding](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `None`

**Comment by GeoTox** [comment]
So what's your advice? Should I enable multi label detection in detect.py and then select a prediction with a higher confidence?

**Comment by glenn-jocher** [comment]
@GeoTox my advice is to update your code. Up to date code will not produce results as you've shown.

**Comment by GeoTox** [comment]
@glenn-jocher thanks, will try!

**Comment by GeoTox** [comment]
UPD: I tried new code, but the problem is still there. I use torch hub to load the model. What should my next steps be? I want also to try to train the model on padded images, not resized, what do you think?

**Comment by glenn-jocher** [comment]
@GeoTox current master will not produce multiple labels on the same box when using detect.py or pytorch hub models. If you believe you have a reproducible issue then I would recommend you file a bug report, providing a minimum reproducible example.

We've created a few short guidelines below to help users provide what we need in order to get started investigating a possible problem.

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/). Your code that reproduces the problem should be:

* ✅ **Minimal** – Use as little code as possible that still produces the same problem
* ✅ **Complete** – Provide **all** parts someone else needs to reproduce your problem in the question itself
* ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

In addition to the above requirements, for [Ultralytics](https://ultralytics.com/) to provide assistance your code should be:

* ✅ **Current** – Verify that your code is up-to-date with current GitHub [master](https://github.com/ultralytics/yolov5/tree/master), and if necessary `git pull` or `git clone` a new copy to ensure your problem has not already been resolved by previous commits.
* ✅ **Unmodified** – Your problem must be reproducible without any modifications to the codebase in this repository. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all of the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/yolov5/issues/new/choose) and providing a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us better understand and diagnose your problem. 

Thank you! 😃

**Comment by GeoTox** [comment]
Let me try to re-train my model using updated code and if the problem still exists, I'll post a bug report.

**Comment by glenn-jocher** [comment]
@GeoTox no you don't need to retrain your model, `multi_label` argument is default False in current master, therefore each box can only have one unique label when called by detect.py or pytorch hub.
https://github.com/ultralytics/yolov5/blob/e8c52374035fd2fb5a0b0029eaa5e5705186df17/utils/general.py#L524-L525

**Comment by GeoTox** [comment]
@glenn-jocher it's the same for the hub models, right? Is the multi_label argument the only reason (probably not) of such behaviour? Let me re-phrase my problem. So, when predicting symbols on images, the model predicts all of them correctly, but sometimes it just repeats one of the boxes and places it on top of the other one. So, let's say the model correctly predicted the following letters (boxes) on an image: 'A', 'B', 'C'. It then might create another 'B' box and place it on top of the 'A' box prediction.

**Comment by zxsitu** [comment]
@GeoTox @glenn-jocher What is the format for creating multi-label datasets?
For example
```
class1 x_center y_center width height (one space) class2 x_center y_center width height
```
or
```
class1 x_center y_center width height
class2 x_center y_center width height
```
Also, in `dataset.yaml`, is it ok to add categories like this normally `names: [class1, class2]`?

**Comment by GeoTox** [comment]
@ilem777 as I know, you should use the second method.

```
class1 x_center y_center width height
class2 x_center y_center width height
```
i.e. multiple lines with same box params but different classes.

**Comment by zxsitu** [comment]
Got it, thank you!
